### PR TITLE
feat: add JWT implementation

### PR DESCRIPTION
**Pull request by MartinOnufrak**
This pull request introduces the getJWTToken method to the internet module.
Changes
•	New Method: getJWTToken
•	Purpose: Generates a JWT token.
•	Parameters:
•	std::optional<std::map<std::string, std::string>> header: The header to be included in the JWT.
•	std::optional<std::map<std::string, std::string>> payload: The payload to be included in the JWT.
•	std::optional<std::string> refDate: The refference date to be included in the JWT.
•	Returns: A std::string representing the generated JWT token.

The implementation using few other methods:
•	getJWTAlgorithm
•	toJSON
•	toBase64UrlEncode

### CODE REVIEW

File: src/modules/internet.cpp [hunk 1]
```diff
@@ -348,4 +352,89 @@ std::string anonymousUsername(unsigned maxLength)
...
+                   {"jti", faker::string::uuid()}};
+    }
+
+    std::string headerToJSON = toJSON(header.value());
+    std::string encodedHeader = toBase64UrlEncode(headerToJSON);
+
+    std::string payloadToJSON = toJSON(payload.value());
+    std::string encodedPayload = toBase64UrlEncode(payloadToJSON);
+
+    std::string signature = faker::string::alphanumeric(64);
```
Comment (cieslarmichal) [inline src/modules/internet.cpp:435]: consider using `const auto` instead std::string when its possible

File: src/modules/internet.cpp [hunk 2]
```diff
@@ -348,4 +352,89 @@ std::string anonymousUsername(unsigned maxLength)
...
+
+    // maybe add option to set ref date to date functions then refactor this
+    std::string iatDefault = faker::date::recentDate(faker::date::dayOfMonth(), faker::date::DateFormat::Timestamp);
+    std::string expDefault = faker::date::soonDate(faker::date::dayOfMonth(), faker::date::DateFormat::Timestamp);
+    std::string nbfDefault = faker::date::anytime(faker::date::DateFormat::Timestamp);
+
+    std::string algorithm(getJWTAlgorithm());
+
+    if (!header)
+        header = {{"alg", algorithm}, {"typ", "JWT"}};
```
Comment (cieslarmichal) [inline src/modules/internet.cpp:417]: please add parentheses `{}` - always for readability

File: include/faker-cxx/internet.h [hunk 3]
```diff
@@ -350,4 +352,74 @@ FAKER_CXX_EXPORT std::string_view domainSuffix();
...
+
+/**
+ * @brief Returns the algorithm used for signing the JWT.
+ *
+ * This function provides the algorithm that is used to sign the JSON Web Token (JWT).
+ *
+ * @returns A string view representing the JWT signing algorithm.
+ *
+ * @code
+ * std::string_view algorithm = faker::internet::getJWTAlgorithm(); // "HS256"
```
Comment (cieslarmichal) [inline include/faker-cxx/internet.h:387]: we don't need to assign it to variable in examples, just `faker::internet::getJWTAlgorithm(); // "HS256"`

Comment (cieslarmichal) [reply include/faker-cxx/internet.h:387]: apply to other functions

File: include/faker-cxx/internet.h [hunk 4]
```diff
@@ -350,4 +352,74 @@ FAKER_CXX_EXPORT std::string_view domainSuffix();
...
+ * @param data The map containing key-value pairs to be converted to JSON.
+ *
+ * @returns A JSON formatted string representing the input map.
+ *
+ * @code
+ * std::map<std::string, std::string> data = {{"name", "John"}, {"age", "30"}};
+ * std::string json = faker::internet::toJSON(data);  // json is now "{\"name\":\"John\",\"age\":\"30\"}"
+ * @endcode
+ */
+FAKER_CXX_EXPORT std::string toJSON(std::map<std::string, std::string>& data);
```
Comment (cieslarmichal) [inline include/faker-cxx/internet.h:423]: arg could be const ref

File: include/faker-cxx/internet.h [hunk 5]
```diff
@@ -350,4 +352,74 @@ FAKER_CXX_EXPORT std::string_view domainSuffix();
...
+ * @param input The string to be encoded.
+ *
+ * @returns A Base64 URL encoded string.
+ *
+ * @code
+ * std::string input = "Hello, World!";
+ * std::string encoded = faker::internet::toBase64UrlEncode(input); // "SGVsbG8sIFdvcmxkIQ"
+ * @endcode
+ */
+FAKER_CXX_EXPORT std::string toBase64UrlEncode(std::string& input);
```
Comment (cieslarmichal) [inline include/faker-cxx/internet.h:407]: arg could be const ref

File: include/faker-cxx/internet.h [hunk 6]
```diff
@@ -350,4 +352,74 @@ FAKER_CXX_EXPORT std::string_view domainSuffix();
...
+ *
+ * @code
+ * std::map<std::string, std::string> header = {{"alg", "HS256"}, {"typ", "JWT"}};
+ * std::map<std::string, std::string> payload = {{"sub", "1234567890"}, {"name", "John Doe"}, {"admin", "true"}};
+ * std::string token = faker::internet::getJWTToken(header, payload); // "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9..."
+ * @endcode
+ */
+FAKER_CXX_EXPORT std::string getJWTToken(std::optional<std::map<std::string, std::string>> header = std::nullopt,
+                                         std::optional<std::map<std::string, std::string>> payload = std::nullopt,
+                                         std::optional<std::string> refDate = std::nullopt);
```
Comment (cieslarmichal) [inline include/faker-cxx/internet.h:377]: try to adding const ref to arguments

File: src/modules/internet.cpp [hunk 7]
```diff
@@ -348,4 +352,89 @@ std::string anonymousUsername(unsigned maxLength)
...
+        {
+            encodedInput.push_back(base64Chars[(value >> validBits) & 0x3F]);
+            validBits -= 6;
+        }
+    }
+
+    if (validBits > -6)
+        encodedInput.push_back(base64Chars[((value << 8) >> validBits) & 0x3F]);
+    while (encodedInput.size() % 4)
+        encodedInput.push_back('=');
```
Comment (cieslarmichal) [inline src/modules/internet.cpp:377]: please add parentheses

File: src/modules/internet.cpp [hunk 8]
```diff
@@ -348,4 +352,89 @@ std::string anonymousUsername(unsigned maxLength)
...
+    return encodedInput;
+}
+
+std::string toJSON(std::map<std::string, std::string>& data)
+{
+    std::string json = "{";
+    for (auto it = data.begin(); it != data.end(); ++it)
+    {
+        if (it != data.begin())
+            json += ",";
```
Comment (cieslarmichal) [inline src/modules/internet.cpp:392]: please add parentheses

File: tests/modules/internet_test.cpp [hunk 9]
```diff
@@ -803,3 +804,9 @@ TEST_F(InternetTest, shouldGenerateAnonymousUsernameWithMaxLength)
 
     ASSERT_EQ(generatedUsername.length(), 20);
 }
+
```
Comment (cieslarmichal) [inline tests/modules/internet_test.cpp:807]: please add more tests, each public function should be tested, in jwt function try to test different parameters also

**Review by cieslarmichal — CHANGES_REQUESTED** [review]
Good job, couple minor comments, and please merge latest changes into your branch.
It would be nice to have more tests here

File: tests/modules/internet_test.cpp (hunk 9 above)
Comment (MartinOnufrak) [reply tests/modules/internet_test.cpp:807]: sure so the helper functions coud stay puplic ?

Comment (cieslarmichal) [reply tests/modules/internet_test.cpp:807]: I think they could be moved to anonymous namespace in cpp file as they are not public API

**Comment by MartinOnufrak** [comment]
All the requirements should be fullfilled

File: include/faker-cxx/internet.h [hunk 10]
```diff
@@ -350,4 +352,77 @@ FAKER_CXX_EXPORT std::string_view domainSuffix();
...
+ * @param data The map containing key-value pairs to be converted to JSON.
+ *
+ * @returns A JSON formatted string representing the input map.
+ *
+ * @code
+ * std::map<std::string, std::string> data = {{"name", "John"}, {"age", "30"}};
+ * faker::internet::toJSON(data);  // json is now "{\"name\":\"John\",\"age\":\"30\"}"
+ * @endcode
+ */
+std::string toJSON(std::map<std::string, std::string>& data);
```
Comment (cieslarmichal) [inline include/faker-cxx/internet.h:427]: move to internet.cpp

**Review by cieslarmichal — CHANGES_REQUESTED** [review]
Please merge latest changes from main branch and move utility functions from include file

File: include/faker-cxx/internet.h (hunk 10 above)
Comment (MartinOnufrak) [reply include/faker-cxx/internet.h:427]: then tests for the util methods wouldnt work if definition wouldnt be in header

Comment (cieslarmichal) [reply include/faker-cxx/internet.h:427]: yup, we don't need them (my previous comment was if they were in include but I think its better to keep them outside API file)

Comment (cieslarmichal) [reply include/faker-cxx/internet.h:427]: just delete tests, and lets focus on testing main jwt function

**Review by cieslarmichal — CHANGES_REQUESTED** [review]
Your changes include some submodule related stuff, try to git aubmodule update to get latest we have in repo

**Comment by cieslarmichal** [comment]
Build fails, please fix it

**Comment by MartinOnufrak** [comment]
sorry when i updated submodule it buold up to over 120 errors along the whole project i didnt expect that

File: src/modules/internet.cpp [hunk 11]
```diff
@@ -22,6 +26,89 @@
 #include "internet_data.h"
 #include "modules/string_data.h"
 
+namespace faker::internet::utility
```
Comment (cieslarmichal) [inline src/modules/internet.cpp:29]: We can use anonymous namespace here so namespace { without name

**Review by cieslarmichal — APPROVED** [review]

**Review by cieslarmichal — APPROVED** [review]

**Pull request merged**